llvm-external-projects/iree-dialects/lib/Dialect/VectorExt/IR/VectorExtDialect.cpp

```diff
@@ -0,0 +1,90 @@
+// Copyright 2023 The IREE Authors
+//
+// Licensed under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+
+#include "iree-dialects/Dialect/VectorExt/IR/VectorExtDialect.h"
+#include "iree-dialects/Dialect/VectorExt/IR/VectorExtOps.h"
+#include "mlir/IR/DialectImplementation.h"
+#include "llvm/ADT/TypeSwitch.h"
+
+using namespace mlir;
+using namespace mlir::iree_compiler::IREE::VectorExt;
+
+#include "iree-dialects/Dialect/VectorExt/IR/VectorExtEnums.cpp.inc" // IWYU pragma: keep
+
+#define GET_ATTRDEF_CLASSES
+#include "iree-dialects/Dialect/VectorExt/IR/VectorExtAttrs.cpp.inc" // IWYU pragma: keep
+
+void IREEVectorExtDialect::initialize() {
+
+  addAttributes<
+#define GET_ATTRDEF_LIST
+#include "iree-dialects/Dialect/VectorExt/IR/VectorExtAttrs.cpp.inc"
+      >();
+
+#define GET_OP_LIST
+  addOperations<
+#include "iree-dialects/Dialect/VectorExt/IR/VectorExtOps.cpp.inc"
+      >();
+}
+
+#include "iree-dialects/Dialect/VectorExt/IR/VectorExtDialect.cpp.inc"
+
+// Parses an attribute with syntax
+// <"BatchX"<"VecX", 2>, 4>
+Attribute PerDimLayoutAttr::parse(AsmParser &parser, Type type) {
```

Comment:
Cant this be done using assembly format? Will be much easier to maintain?

Reply:
This is a good question. I'm not sure. So I have a vector of labels and a vector of shapes. We don't know how many there are ahead of time, so can this be done using assembly format?

Reply:
There should be a assembly format handling for array attribute types.

Reply:
Thanks let me look into it.

Reply:
Okay so seems like you can print out the array attributes but we won't be able to get it in this form. It will probably be something like ["BatchX", "VecX"], [4, 2] (which is actually similar to the MLIR PR I had earlier). While they both convey the same information,  the reason I chose this syntax is it more closely matches what the layout is doing. So for example, first we group the tensor dimension (such as row/col) into groups of 2 and call it VecX, then we group the VecX dimension into groups of 4 and call it BatchX. That being said, if people find ["BatchX", "VecX"], [4, 2] to be clearer I am open to changing it.

Reply:
It would look like this 

```
module {
  func.func @specify_layout(%arg0: memref<32x32xf16>, %arg1: memref<32x32xf16>) -> vector<32x32xf16> {
    %cst = arith.constant 0.000000e+00 : f16
    %c0 = arith.constant 0 : index
    %0 = vector.transfer_read %arg0[%c0, %c0], %cst {in_bounds = [true, true]} : memref<32x32xf16>, vector<32x32xf16>
    %1 = iree_vector_ext.layout_conflict_resolution %0 {desiredLayout = #iree_vector_ext.layout<<[BatchY, LaneY, VecX], [4, 2, 4]>, <[BatchX, LaneX, VecY], [2, 4, 4]>>, sourceLayout = #iree_vector_ext.layout<<[BatchX, LaneX, VecY], [2, 4, 4]>, <[BatchY, LaneY, VecX], [4, 2, 4]>>} : vector<32x32xf16> -> vector<32x32xf16>
    return %1 : vector<32x32xf16>
  }
}
```

Reply:
Upto you really. Using assembly format means less maintenance overhead. But if you want a custom parser printer then that's your call.

Reply:
My preference would be to keep it for now, but let's keep an eye on the maintenance overhead and if it seems like that's getting overwhelming we can go to the alternative representation.